File: typescript/packages/core/src/commands/builtin/trello/reads.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { TrelloAccessor } from '../../../accessor/trello.ts'
import {
  getBoard,
  getCard,
  listBoardLabels,
  listBoardLists,
  listBoardMembers,
  listCardComments,
  listListCards,
  listWorkspaceBoards,
  listWorkspaces,
} from '../../../core/trello/_client.ts'
import {
  normalizeBoard,
  normalizeCard,
  normalizeComment,
  normalizeLabel,
  normalizeList,
  normalizeMember,
  toJsonBytes,
} from '../../../core/trello/normalize.ts'
import { IOResult } from '../../../io/types.ts'
import { ResourceName, type PathSpec } from '../../../types.ts'
import {
  command,
  type CommandFnResult,
  type CommandOpts,
  type RegisteredCommand,
} from '../../config.ts'
import { CommandSpec, OperandKind, Operand } from '../../spec/types.ts'

type Runner = (accessor: TrelloAccessor, texts: string[]) => Promise<Uint8Array>

interface TrelloRead {
  name: string
  runner: Runner
  spec: CommandSpec
}

const SPEC_NONE = new CommandSpec({})
const SPEC_ARG = new CommandSpec({ rest: new Operand({ kind: OperandKind.TEXT }) })

function first(texts: string[], label: string): string {
  const value = texts[0]
  if (value === undefined || value === '') throw new Error(`${label} is required`)
  return value
}

async function runBoardList(accessor: TrelloAccessor): Promise<Uint8Array> {
  const boards: unknown[] = []
  for (const workspace of await listWorkspaces(accessor.transport)) {
    const workspaceId = typeof workspace.id === 'string' ? workspace.id : ''
    for (const board of await listWorkspaceBoards(accessor.transport, workspaceId)) {
      boards.push(normalizeBoard(board))
    }
  }
  return toJsonBytes(boards)
}

async function runBoardShow(accessor: TrelloAccessor, texts: string[]): Promise<Uint8Array> {
  const board = await getBoard(accessor.transport, first(texts, 'board id'))
  return toJsonBytes(normalizeBoard(board))
}

async function runBoardMembers(accessor: TrelloAccessor, texts: string[]): Promise<Uint8Array> {
  const members = await listBoardMembers(accessor.transport, first(texts, 'board id'))
  return toJsonBytes(members.map((member) => normalizeMember(member)))
}

async function runListList(accessor: TrelloAccessor, texts: string[]): Promise<Uint8Array> {
  const lists = await listBoardLists(accessor.transport, first(texts, 'board id'))
  return toJsonBytes(lists.map((lst) => normalizeList(lst)))
}

async function runLabelList(accessor: TrelloAccessor, texts: string[]): Promise<Uint8Array> {
  const labels = await listBoardLabels(accessor.transport, first(texts, 'board id'))
  return toJsonBytes(labels.map((label) => normalizeLabel(label)))
}

async function runCardList(accessor: TrelloAccessor, texts: string[]): Promise<Uint8Array> {
  const cards = await listListCards(accessor.transport, first(texts, 'list id'))
  return toJsonBytes(cards.map((card) => normalizeCard(card)))
}

async function runCardShow(accessor: TrelloAccessor, texts: string[]): Promise<Uint8Array> {
  const card = await getCard(accessor.transport, first(texts, 'card id'))
  return toJsonBytes(normalizeCard(card))
}

async function runCardComments(accessor: TrelloAccessor, texts: string[]): Promise<Uint8Array> {
  const cardId = first(texts, 'card id')
  const comments = await listCardComments(accessor.transport, cardId)
  return toJsonBytes(comments.map((comment) => normalizeComment(comment, cardId)))
}

const TRELLO_READS: readonly TrelloRead[] = [
  { name: 'trello board list', runner: (a) => runBoardList(a), spec: SPEC_NONE },
  { name: 'trello board show', runner: (a, t) => runBoardShow(a, t), spec: SPEC_ARG },
  { name: 'trello board members', runner: (a, t) => runBoardMembers(a, t), spec: SPEC_ARG },
  { name: 'trello list list', runner: (a, t) => runListList(a, t), spec: SPEC_ARG },
  { name: 'trello label list', runner: (a, t) => runLabelList(a, t), spec: SPEC_ARG },
  { name: 'trello card list', runner: (a, t) => runCardList(a, t), spec: SPEC_ARG },
  { name: 'trello card show', runner: (a, t) => runCardShow(a, t), spec: SPEC_ARG },
  { name: 'trello card comments', runner: (a, t) => runCardComments(a, t), spec: SPEC_ARG },
]

export function makeTrelloReadCommands(): RegisteredCommand[] {
  const commands: RegisteredCommand[] = []
  for (const entry of TRELLO_READS) {
    commands.push(
      ...command({
        name: entry.name,
        resource: ResourceName.TRELLO,
        spec: entry.spec,
        fn: async (
          accessor,
          _paths: PathSpec[],
          texts: string[],
          _opts: CommandOpts,
        ): Promise<CommandFnResult> => {
          const data = await entry.runner(accessor as TrelloAccessor, texts)
          return [data, new IOResult()]
        },
      }),
    )
  }
  return commands
}

File: python/mirage/commands/builtin/trello/reads.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import functools
from collections.abc import Awaitable, Callable
from dataclasses import dataclass

from mirage.accessor.trello import TrelloAccessor
from mirage.commands.registry import command
from mirage.commands.spec.types import (CommandSpec, FlagView, Operand,
                                        OperandKind)
from mirage.core.trello._client import (get_board, get_card, list_board_labels,
                                        list_board_lists, list_board_members,
                                        list_card_comments, list_list_cards,
                                        list_workspace_boards, list_workspaces)
from mirage.core.trello.normalize import (normalize_board, normalize_card,
                                          normalize_comment, normalize_label,
                                          normalize_list, normalize_member,
                                          to_json_bytes)
from mirage.io.stream import yield_bytes
from mirage.io.types import ByteSource, IOResult
from mirage.types import PathSpec

Runner = Callable[[TrelloAccessor, list[str], FlagView], Awaitable[bytes]]


@dataclass(frozen=True, slots=True)
class TrelloRead:
    name: str
    runner: Runner
    spec: CommandSpec


SPEC_NONE = CommandSpec()
SPEC_ARG = CommandSpec(rest=Operand(kind=OperandKind.TEXT))


def _first(texts: list[str], label: str) -> str:
    if not texts:
        raise ValueError(f"{label} is required")
    return texts[0]


async def _run_board_list(accessor: TrelloAccessor, texts: list[str],
                          fl: FlagView) -> bytes:
    config = accessor.config
    boards: list[dict[str, object]] = []
    for workspace in await list_workspaces(config):
        for board in await list_workspace_boards(config, workspace["id"]):
            boards.append(normalize_board(board))
    return to_json_bytes(boards)


async def _run_board_show(accessor: TrelloAccessor, texts: list[str],
                          fl: FlagView) -> bytes:
    board = await get_board(accessor.config, _first(texts, "board id"))
    return to_json_bytes(normalize_board(board))


async def _run_board_members(accessor: TrelloAccessor, texts: list[str],
                             fl: FlagView) -> bytes:
    config = accessor.config
    members = await list_board_members(config, _first(texts, "board id"))
    return to_json_bytes([normalize_member(member) for member in members])


async def _run_list_list(accessor: TrelloAccessor, texts: list[str],
                         fl: FlagView) -> bytes:
    config = accessor.config
    lists = await list_board_lists(config, _first(texts, "board id"))
    return to_json_bytes([normalize_list(lst) for lst in lists])


async def _run_label_list(accessor: TrelloAccessor, texts: list[str],
                          fl: FlagView) -> bytes:
    config = accessor.config
    labels = await list_board_labels(config, _first(texts, "board id"))
    return to_json_bytes([normalize_label(label) for label in labels])


async def _run_card_list(accessor: TrelloAccessor, texts: list[str],
                         fl: FlagView) -> bytes:
    config = accessor.config
    cards = await list_list_cards(config, _first(texts, "list id"))
    return to_json_bytes([normalize_card(card) for card in cards])


async def _run_card_show(accessor: TrelloAccessor, texts: list[str],
                         fl: FlagView) -> bytes:
    card = await get_card(accessor.config, _first(texts, "card id"))
    return to_json_bytes(normalize_card(card))


async def _run_card_comments(accessor: TrelloAccessor, texts: list[str],
                             fl: FlagView) -> bytes:
    config = accessor.config
    card_id = _first(texts, "card id")
    comments = await list_card_comments(config, card_id)
    return to_json_bytes(
        [normalize_comment(comment, card_id=card_id) for comment in comments])


TRELLO_READS: tuple[TrelloRead, ...] = (
    TrelloRead("trello board list", _run_board_list, SPEC_NONE),
    TrelloRead("trello board show", _run_board_show, SPEC_ARG),
    TrelloRead("trello board members", _run_board_members, SPEC_ARG),
    TrelloRead("trello list list", _run_list_list, SPEC_ARG),
    TrelloRead("trello label list", _run_label_list, SPEC_ARG),
    TrelloRead("trello card list", _run_card_list, SPEC_ARG),
    TrelloRead("trello card show", _run_card_show, SPEC_ARG),
    TrelloRead("trello card comments", _run_card_comments, SPEC_ARG),
)


async def _dispatch(
    entry: TrelloRead,
    accessor: TrelloAccessor,
    paths: list[PathSpec],
    *texts: str,
    **flags: object,
) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(flags, spec=entry.spec)
    data = await entry.runner(accessor, list(texts), fl)
    return yield_bytes(data), IOResult()


def make_trello_read_commands() -> list[Callable[..., object]]:
    commands: list[Callable[..., object]] = []
    for entry in TRELLO_READS:
        commands.append(
            command(entry.name, resource="trello",
                    spec=entry.spec)(functools.partial(_dispatch, entry)))
    return commands
